### PR TITLE
Edit command unique feature

### LINKED ISSUE #116: Update UG for edit by identifier command

### PR DESCRIPTION
**Pull request by matthewken19**
Resolves #116, #109

- Update EditCommandParser to parse unique identifier as well

- Add CLI Syntax for unique identifier for edit command

- Update EditCommand to support editing student based on unique identifier aside from index

- Add size method for ArgumentMultimap class

- Add JUnit Tests for EditCommand and EditCommandParser

- Update UG

### CODE REVIEW

File: src/main/java/educonnect/logic/commands/EditCommand.java [hunk 1]
```diff
@@ -40,10 +45,14 @@ public class EditCommand extends Command {
     public static final String COMMAND_WORD = "edit";
 
     public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits the details of the student identified "
-            + "by the index number used in the displayed student list. "
+            + "by the identifier used in the displayed student list. "
             + "Existing values will be overwritten by the input values.\n\n"
 
-            + "Parameters: INDEX (must be a positive integer)\n"
+            + "Parameters: <choose onl 1>"
```
Comment (darryl-chan) [inline src/main/java/educonnect/logic/commands/EditCommand.java:51]: Mispelling here

**Comment by darryl-chan** [comment]
Fix merge conflicts and mispelling. Otherwise LGTM

File: src/main/java/educonnect/logic/parser/EditCommandParser.java [hunk 2]
```diff
@@ -41,55 +51,80 @@ public class EditCommandParser implements Parser<EditCommand> {
...
+        if (argMultimap.size() < 2) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, EditCommand.MESSAGE_NOT_EDITED));
+        }
+        // check for duplicate field
+        argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_NAME, PREFIX_STUDENT_ID, PREFIX_EMAIL, PREFIX_TELEGRAM_HANDLE,
+            PREFIX_LINK, PREFIX_TIMETABLE);
+        // check for duplicate identifier
+        identifierArgMultimap.verifyNoDuplicatePrefixesFor(EDIT_ID_PREFIX_EMAIL, EDIT_ID_PREFIX_STUDENT_ID,
+            EDIT_ID_PREFIX_INDEX, EDIT_ID_PREFIX_TELEGRAM_HANDLE);
+        // check for invalid preamble
```
Comment (tanjiajiajun) [inline src/main/java/educonnect/logic/parser/EditCommandParser.java:80]: Is there a check to see if we use muitiple prefix?
eg edit s: e: .....

File: src/main/java/educonnect/logic/commands/EditCommand.java [hunk 3]
```diff
@@ -62,35 +71,48 @@ public class EditCommand extends Command {
...
         List<Student> lastShownList = model.getFilteredStudentList();
-
+        // no student matches the unique identifier
+        if (lastShownList.isEmpty()) {
+            throw new CommandException(MESSAGE_NO_STUDENT_FOUND);
+        }
+        // index out of bound
         if (index.getZeroBased() >= lastShownList.size()) {
-            throw new CommandException(Messages.MESSAGE_INVALID_STUDENT_DISPLAYED_INDEX);
+            throw new CommandException(String.format(MESSAGE_INDEX_OUT_OF_BOUNDS, lastShownList.size()));
```
Comment (tanjiajiajun) [inline src/main/java/educonnect/logic/commands/EditCommand.java:115]: Thanks for fixing this bug :)

File: src/test/java/educonnect/logic/parser/EditCommandParserTest.java [hunk 4]
```diff
@@ -56,143 +66,204 @@ public class EditCommandParserTest {
...
+        assertParseFailure(parser, "i:1" + INVALID_STUDENT_ID_DESC + EMAIL_DESC_AMY, StudentId.MESSAGE_CONSTRAINTS);
 
         // while parsing {@code PREFIX_TAG} alone will reset the tags of the {@code Student} being edited,
         // parsing it together with a valid tag results in error
-        assertParseFailure(parser, "1" + TAG_DESC_FRIEND + TAG_DESC_HUSBAND + TAG_EMPTY, Tag.MESSAGE_CONSTRAINTS);
-        assertParseFailure(parser, "1" + TAG_DESC_FRIEND + TAG_EMPTY + TAG_DESC_HUSBAND, Tag.MESSAGE_CONSTRAINTS);
-        assertParseFailure(parser, "1" + TAG_EMPTY + TAG_DESC_FRIEND + TAG_DESC_HUSBAND, Tag.MESSAGE_CONSTRAINTS);
+        assertParseFailure(parser, "i:1" + TAG_DESC_FRIEND + TAG_DESC_HUSBAND + TAG_EMPTY, Tag.MESSAGE_CONSTRAINTS);
+        assertParseFailure(parser, "i:1" + TAG_DESC_FRIEND + TAG_EMPTY + TAG_DESC_HUSBAND, Tag.MESSAGE_CONSTRAINTS);
+        assertParseFailure(parser, "i:1" + TAG_EMPTY + TAG_DESC_FRIEND + TAG_DESC_HUSBAND, Tag.MESSAGE_CONSTRAINTS);
```
Comment (tanjiajiajun) [inline src/test/java/educonnect/logic/parser/EditCommandParserTest.java:117]: Test coverage mostly done on i:, maybe can use other prefixes

File: src/main/java/educonnect/logic/parser/EditCommandParser.java (hunk 2 above)
Comment (matthewken19) [reply src/main/java/educonnect/logic/parser/EditCommandParser.java:80]: it is not allowed. The code before/after that checks for multiple prefix and raise an error.

File: src/main/java/educonnect/logic/commands/EditCommand.java (hunk 1 above)
Comment (matthewken19) [reply src/main/java/educonnect/logic/commands/EditCommand.java:51]: Good catch. Thanks for the input.

**Pull request merged**